pkg/testgen/context.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testgen
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// TestContext holds contextual information used to configure a test, such as a mapping
+// of which branch names map to which TestGrid dashboard names
+type TestContext struct {
+	Branches []string
+
+	// PresubmitDashboardName is the name of the TestGrid dashboard to which presubmit tests should
+	// be added. If unset, no TestGrid annotations will be added to presubmit tests.
+	PresubmitDashboardName string
+
+	// PeriodicDashboardName is the name of the TestGrid dashboard to which periodic tests should
+	// be added. If unset, no TestGrid annotations will be added to periodic tests.
+	PeriodicDashboardName string
+
+	Org  string
+	Repo string
+
+	// Descriptor is a string which, if set, is inserted into periodic test names.
+	// An example would be "previous" which would result in "my-test" having the name
+	// "ci-<repo>-previous-my-test", where the test would've been called
+	// "ci-<repo>-my-test" if the descriptor was not set
+	Descriptor string
+
+	presubmits []*PresubmitTest
+	periodics  []*PeriodicTest
+
+	minutesCounter time.Time
+}
+
+// RequiredPresubmit adds a presubmit which is run by default and required to pass before a PR can be merged
+func (tc *TestContext) RequiredPresubmit(test *Test) {
+	tc.addPresubmit(test, true, false)
+}
+
+// RequiredPresubmit adds a presubmit which is not run by default and is optional
+func (tc *TestContext) OptionalPresubmit(test *Test) {
+	tc.addPresubmit(test, false, true)
+}
+
+func (tc *TestContext) addPresubmit(test *Test, alwaysRun bool, optional bool) {
+	test.Name = tc.presubmitTestName(test.Name)
+
+	if tc.PresubmitDashboardName != "" {
+		addTestGridAnnotations(tc.PresubmitDashboardName)(test)
+	}
+
+	tc.presubmits = append(tc.presubmits, &PresubmitTest{
+		Test:      *test,
+		Branches:  tc.Branches,
+		AlwaysRun: alwaysRun,
+		Optional:  optional,
+	})
+}
+
+// Periodic adds periodic tests which will run every `periodicityHours` hours, at some minute
+// within the hour, one test for each configured branch
+func (tc *TestContext) Periodics(test *Test, periodicityHours int) {
+	for _, branch := range tc.Branches {
+		test.Name = tc.periodicTestName(test.Name)
+
+		if tc.PeriodicDashboardName != "" {
+			addTestGridAnnotations(tc.PeriodicDashboardName)(test)
+		}
```

Comment:
I don't fully understand this condition block. Why do we only add annotations if the dashboard name is defined?